### PR TITLE
Draft: non-linear CHC parsing

### PR DESCRIPTION
**Pull request by gruhn**
Ignore my changes in all non cpp/hpp files (Dockerfile, docker-compose.yaml, etc.). I'll remove them before merging.

I mainly want feedback on some questions, that I have written down as inline comments marked with `QUESTION`.

### CODE REVIEW

**Comment by ffrohn** [comment]
Please do not use dashs in filenames for consistency.

> QUESTION: first I wanted to give `nonLinearRules` the same type
> as `rules`, ie `std::map<TransIdx, Rule>`, but as I understood,
> we won't "add" the non linear transitions to the ITS, right?

I'd say we add them to the ITS, we just store linear and
non-linear rules separately. So it would make sense to have
dedicated indices for non-linear rules. But if we don't need
them (since we can, e.g., pass references around instead) it's
also fine not to have indices for non-linear rules.

> We can't even re-use the `TransIdx` when we eventually derive
> linear rules from the non linear ones, since they are not
> unique, right?

So far, every original rule and every learned rule has a unique
index. If we stick to that, then every original non-linear rule
would have an index, and every rule that is derived from such a
rule (no matter if it's non-linear or linear) would have a unique
index, too.

> I mean we can derive different linear rules from the non linear
> rules and then they would need different `TransIdx`, right?

Yes, that's right.

> QUESTION: all argument variables are pairwise disjoint even in
> the non-linear case, right?

Yes, I hope so. If they aren't, we should adapt that.

> QUESTION: what exactly does `holds_alternative` do and how does
> it work?

An object o of type std::variant<A, B> can store either an
instance of A or an instance of B. Then holds_alternative<A>(o)
evaluates to true if and only if o stores an instance of A.

https://en.cppreference.com/w/cpp/utility/variant/holds_alternative

> QUESTION: what is the meaning of `Rel::buildEq(NumVar::loc_var,
> c.lhs.loc)` again?  So let's say we have linear CHC like:
>
>         F(X) /\ X > 1   ==>   G(X)
>
> then we encode the predicate "names" (ie. F and G) with
> `LocationIdx`, right?

Yes, and LocationIdx is just a number. So for example, we could
represent F with 1 and G with 2.

> So `c.lhs.loc` basically represents `F`.

Right. Then c.lhs.loc would be 1.

> And `NumVar::loc_var` is a constant that's always 1 or what?

No, it's a variable that stores the current location during a run
of the ITS.

> And then we extend the guard with:
>
>       X > 1 /\ F == 1

No, with

> X > 1 /\ 1 == loc_var

as 1 represents F.

**Comment by gruhn** [comment]
> No, it's a variable that stores the current location during a run of the ITS.

Ah ok. So we don't keep track of the source nodes of each transition with dedicated fields on the `Rule` class, but we kind of reify that information using the guard? Basically all transitions are possible from all nodes but the guard is unsatisfiable if the current node does not match, right? That's a bit surprising to me. Just finding allI transitions from the current node is just a lookup but by using the guard we offload that work to the SMT solver right?

**Comment by ffrohn** [comment]
Yes, but to work around that, we use the [dependency graph](https://github.com/LoAT-developers/LoAT/blob/master/src/its/dependencygraph.hpp). It's a graph whose nodes are transitions, and there is an edge from t1 to t2 if t2 can be applied after t1. It's pre-computed once in the beginning of the analysis. So we can just use the successors of the transition that was used for previous step as candidate transitions for the next step.

Note that locations are a somewhat arbitrary way to encode the control-flow. No one prevents you from writing, e.g.,

`f(b) -> f(b') [b = 0 /\ b' = 1]`

which is "recursive" if you just consider the locations, but in fact it's not. Thus, the dependency graph is a better way to represent the control-flow.

**Comment by gruhn** [comment]
> I'd say we add them to the ITS, we just store linear and
> non-linear rules separately. So it would make sense to have
> dedicated indices for non-linear rules.

But if we have a non-linear CHC:
```
F   /\    G     ==>    False
```
Then it's technically a sink transition, right? But we wouldn't add it to the set of `sinkTransitions` on the ITS. At least until we can derive a linear transition from it, right? That's what I mean with "adding to the ITS". Intuitively, I would expect that everything of type `TransIdx` corresponds to a transition that "exists" in the ITS. But I guess that's a subjective point.

**Comment by ffrohn** [comment]
Yes, so `sinkTransitions` would only store linear sink transitions, and we would add `F ==> false` or `G ==> false` to `sinkTransitions` as soon as we derived `G` or `F`, respectively. More precisely, the "linear solver" would notify the "non-linear solver" that `G` or `F` has been derived, and then the non-linear solver would notify the linear solver about the new query.

But you are right, we have to be a bit careful about the semantics of the data-structures that store transitions, since it might not always be obvious whether they are for linear transitions, non-linear transitions, or both. As long as we deal with linear or non-linear transitions directly, the difference can be seen on type-level, but `TransIdx` would be the same for both linear and non-linear rules. So maybe it would be better to use `TransIdx` exclusively for linear transitions to avoid bypassing the type system.

**Comment by gruhn** [comment]
> Yes, but to work around that, we use the [dependency graph](https://github.com/LoAT-developers/LoAT/blob/master/src/its/dependencygraph.hpp). It's a graph whose nodes are transitions, and there is an edge from t1 to t2 if t2 can be applied after t1.

I see. But if we can use the dependency graph, why do have put the extra condition in the guard then? Isn't it tautological then during analysis? Do we just need it to compute the dependency graph in the beginning? Then why not keep track of the source node explicitly and build up the dependency graph that way?

> It's pre-computed once in the beginning of the analysis.

Ok so with the nonlinear rules, we have to extend the dependency graph whenever we derive new linear rules, right?

**Comment by ffrohn** [comment]
> I see. But if we can use the dependency graph, why put the extra condition in the guard then? Isn't it tautological then during analysis?

For ADCL we wouldn't need the extra condition, but there are other engines in LoAT (BMC and ABMC) that need it.

> Do we just need it to compute the dependency graph in the beginning?

We need to add nodes to it whenever we learn new transitions.

> Then why not keep track of the source node explicitly and build up the dependency graph that way?

We keep track of the source and target locations of all transitions in `ITSProblem::startAndTargetLocations`, and we use the source and target locations to get an initial approximation of the dependency graph. Then we refine the dependency graph with `ITSProblem::refineDependencyGraph()`.

> Ok so with the nonlinear rules, we have to extend the dependency graph whenever we derive new linear rules, right?

Yes, but we already do that when ADCL learns new clauses. I guess

`ITSProblem::addRule(const Rule &rule, const LocationIdx start)`

should do what you want. It returns the index of the newly added transition. Afterwards, you can invoke

`ITSProblem::refineDependencyGraph(const TransIdx idx)`

to refine the dependency graph locally, i.e., just in the neighborhood of the newly added transition.

**Comment by ffrohn** [comment]
The guard of the chained rule (aka resolvent) is
```
fst.getGuard() & snd.getGuard()->subs(expr::compose(sigma, fst.getUpdate()))
```
where `sigma` is a variable renaming that just affects temporary variables, so it does not affect `loc_var`. Let's say that `fst.getUpdate()` sets `loc_var` to `G`. Then 
```
(loc_var=F)->subs(expr::compose(sigma, fst.getUpdate()))
= (loc_var=F)->fst.getUpdate()     // as sigma does not affect loc_var
= (G=F)                            // as fst.getUpdate() sets loc_var to G
```
So if `G` and `F` are equal, the resulting literal is equivalent to `true`, otherwise it's equivalent to `false`.

**Comment by gruhn** [comment]
Ok, I think it's time to justify some design decisions :grin: 

Remember that we said, the non-linear solver should represent the CHCs simply with the `Clause` struct, [that the parser was outputting already](https://github.com/LoAT-developers/LoAT/pull/1/files#diff-a38a5654d3459fdae8c677a9a390901266ce0df6f511a88fa6b15ce6fa9456d1L54-L61). I extracted `Clause` now into its [own file; turned it into a class; and added some methods](https://github.com/LoAT-developers/LoAT/pull/1/files#diff-e1f7c7e55754d019739056dcf17a4faef02667ffabd789d27fe0535b8c59d35d). 

Now the non-linear solver must be able to restore `Clause` instances from ITS transitions to do resolution with linear CHCs. For that we said the non-linear solver should be able to do that just given a `TransIdx`. I tried to implement this in [this new `clauseFrom` method here](https://github.com/LoAT-developers/LoAT/pull/1/files#diff-b2ecf27763cd072eff20c09b4db5b9ba1fb3f6e8dacbfdf3ad2db2e36fc7d657R31-R52). To restore the LHS predicate, I need to know what the arguments are. As far as I understood, the arguments are always the set of "program variables". However, there's is no implied argument order. I need to pick an argument order. I think it's ok to pick one arbitrarily but I need to do it consistently. So I fixed an order with [attributes on the ITS problem](https://github.com/LoAT-developers/LoAT/pull/1/files#diff-4a0495413456fd40ea011c862d1372896967950631357d94a63fae28e8442dc4R111-R116). I'm assigning these [attributes here during parsing](https://github.com/LoAT-developers/LoAT/pull/1/files#diff-b5e083b73e4622a282fbef6f7fc7f280ef188be3376ff33428ddfdf014b97e53L56-R67). 

There is another reason, I need the program variables on the ITS: When the non-linear solver derives a new linear CHC it needs to convert it into an ITS transition and add it to the ITS problem. For this conversion I need to essentially invoke [this preprocessing logic that is already implemented in the parser](https://github.com/LoAT-developers/LoAT/pull/1/files#diff-b5e083b73e4622a282fbef6f7fc7f280ef188be3376ff33428ddfdf014b97e53L65-L119). So to do that and to avoid redundancy, I extracted this logic and also turned it into [a new ITS method](https://github.com/LoAT-developers/LoAT/pull/1/files#diff-913ba199f34ba9366a9d53e0a7b0f920045f94987e8b32e9a062e0708eb087b9R308-R384). That logic in turn also needs access to the program variables, so that's the second reason why I think it makes sense to add them to the ITS problem class.

**Comment by gruhn** [comment]
So I did some tests with the LIA benchmark. I just did [one round of resolution of all linear clauses with all non-linear clauses and added the resolvents to the ITS](https://github.com/LoAT-developers/LoAT/pull/1/files#diff-b2ecf27763cd072eff20c09b4db5b9ba1fb3f6e8dacbfdf3ad2db2e36fc7d657R57-R90). Apparently, that can prove [UNSAT of quite a few instances already](https://github.com/LoAT-developers/LoAT/pull/1/commits/0f4dbf643bc116e78f30eb17296b1060cabb0c22#diff-34a08e5286823928d5da009f867c3bd81d0b151eb619e662f4a8efe41c4311bbR2-R155) (assuming my implementation is correct). For comparison, I also ran ADCL against the LIA instances by just ignoring all the non-linear CHCs. That wasn't so successful. 

However, I also got segmentation faults on quite a few instances. I dug into it and they occur here in [this preprocessing step](https://github.com/LoAT-developers/LoAT/blob/8719963c469f2b5c37fef036d2d05a558b9ab4b0/src/analysis/preprocessing.cpp#L28). If I understand correctly, in this step we remove transitions that are not "forward" reachable from any of the initial transitions or "backwards" reachable from any of the sink transitions. 

I think what's happening is that sometimes the sink transitions are only reachable via non-linear CHCs. But because they are not accounted for in the dependency graph initially, we falsely remove the sink transitions in the "forward" preprocessing step. And then during "backward" preprocessing, there are no sink transitions anymore, so we [`todo.pop()`](https://github.com/LoAT-developers/LoAT/blob/8719963c469f2b5c37fef036d2d05a558b9ab4b0/src/analysis/preprocessing.cpp#L43) an empty stack.

So I guess we can either skip preprocessing or we extend the dependency graph with all possible transitions, but I don't know what effects this might have. Can we have dependencies, although there are no associated rules in the ITS?

**Comment by ffrohn** [comment]
> I need to pick an argument order. I think it's ok to pick one arbitrarily but I need to do it consistently.

All kinds of variables (`NumVar` and `BoolVar`) implement the operator `<=>`. That gives you a total order on variables, which should be everything you need, right?

> That logic in turn also needs access to the program variables, so that's the second reason why I think it makes sense to add them to the ITS problem class.

I think you need to be able to differentiate between program variables and temporary variables, correct? To do so, you can use `expr::isTempVar` and `expr::isProgVar`.

> So I did some tests with the LIA benchmark [...] For comparison, I also ran ADCL against the LIA instances by just ignoring all the non-linear CHCs. That wasn't so successful.

Sounds great!
 
> However, I also got segmentation faults on quite a few instances. [...] If I understand correctly, in this step we remove transitions that are not "forward" reachable from any of the initial transitions or "backwards" reachable from any of the sink transitions.

Correct.

> I think what's happening is that sometimes the sink transitions are only reachable via non-linear CHCs. But because they are not accounted for in the dependency graph initially, we falsely remove the sink transitions in the "forward" preprocessing step. And then during "backward" preprocessing, there are no sink transitions anymore, so we [`todo.pop()`](https://github.com/LoAT-developers/LoAT/blob/8719963c469f2b5c37fef036d2d05a558b9ab4b0/src/analysis/preprocessing.cpp#L43) an empty stack.

Yes, that makes sense.

> So I guess we can either skip preprocessing or we extend the dependency graph with all possible transitions, but I don't know what effects this might have. Can we have dependencies, although there are no associated rules in the ITS?

From the point of view of the linear solver, we are solving a problem 'incrementally', i.e., we do not know all clauses when the analysis starts. So far, that was not the case -- we always knew all clauses from the very beginning. In such an incremental setting, it makes little sense to remove transitions because they are not reachable, as they might become reachable later on. Thus, we should disable this particular preprocessing step if the problem is non-linear.

However, as the non-linear solver only learns facts, that's only true for forward reachability. Removing clauses that are not backward-reachable from querys / sink transitions should still make sense.

**Comment by ffrohn** [comment]
For the record: The following part of my last post is wrong.

> However, as the non-linear solver only learns facts, that's only true for forward reachability. Removing clauses that are not backward-reachable from querys / sink transitions should still make sense.

The non-linear part learns rules, not facts, so we should disable the pre-processing entirely.

**Comment by gruhn** [comment]
Also just noticed, `chc-LIA_349.smt2` contains a number that is exactly the max value of int32 +1:
```
(or (not O19) (= C14 (div Z13 2147483648)))
```
Parsing gives me an integer overflow. Looks like the parser is storing integers in a normal `int`. I thought `int` automatically uses 64-bits on a 64-bit system but apparently not. Not sure if this can be changed with some compiler flag or something. Refactoring everything to `long` certainly sounds like too much hassle.

**Comment by ffrohn** [comment]
> Also just noticed, `chc-LIA_349.smt2` contains a number that is exactly the max value of int32 +1:
> 
> ```
> line 874:        (or (not O19) (= C14 (div Z13 2147483648)))
> ```
> 
> Parsing gives me an integer overflow. Looks like the parser is storing integers in a normal `int`. I thought `int` automatically uses 64-bits on a 64-bit system but apparently not. Not sure if this can be changed with some compiler flag or something. Refactoring everything to `long` certainly sounds like too much hassle.

That's surprising. Constants are parsed by this line:
```
res.t = Expr(Num(ctx->getText().c_str()));
```
Here, `Num` is an alias for `GiNaC::numeric`, which should be able to store values of arbitrary size. Can you figure out where exactly the overflow happens?

**Comment by gruhn** [comment]
Ah ok! I tracked it down. The problem is specific to div/mod expressions. Here is the offending line:

https://github.com/LoAT-developers/LoAT/blob/termcomp23/src/parser/chc/CHCParseVisitor.cpp#L432

This fixes the issue for me:

https://github.com/LoAT-developers/LoAT/pull/1/commits/4d99e1ace3b63974f158c73f29f73fba6643c264

**Comment by gruhn** [comment]
@ffrohn can you help me understand what `luby` is, in the context of reachability? In particular, this line:

https://github.com/LoAT-developers/LoAT/blob/termcomp23/src/analysis/reachability.cpp#L335

What is the meaning of this pair of integers? Do I have to be careful invoking this function too often? I was planning to restart the linear solver for every linear clause that is added to the ITS. But that would mean that I sometimes restart the solver multiple times in a row, and thereby invoke this `luby_next` function multiple times in a row.

**Comment by ffrohn** [comment]
See [this paper](https://doi.org/10.1016/0020-0190(93)90029-9). It's just a heuristic to determine how often one should perform a restart that works very well for SAT solving, so we adapted it to our setting. If you invoke `luby_next` often due to restarts that are caused by the non-linear solver, then the linear parts will perform fewer restarts by itself on the long run. Maybe it would be better not to call `luby_next` for those restarts, but I think that's not obvious.

**Comment by gruhn** [comment]
FYI: just noticed that for `LIA-Lin_036` ADCL terminates with `unknown` while Z3 (v4.9.1) gives `unsat`. If Z3 is right, ADCL should also be able to proof `unsat` (or timeout) right? But never terminate with `unknown`, right?

**Comment by ffrohn** [comment]
> If Z3 is right, ADCL should also be able to prove `unsat` (or timeout) right?

No. Refutational completeness of ADCL just means that *there is* a run of ADCL that proves unsat. Since ADCL is inherently non-deterministic, that does not imply that the actual run which is performed by the implementation proves unsat.

Moreover, refutational completeness only holds under the assumption that we have a complete SMT solver, a complete redundancy check, and complete acceleration techniques. In practice, all of these things are incomplete.

In other words: Refutational completeness is a property of the calculus, it's not a property of the implementation.

**Comment by gruhn** [comment]
Since rebasing onto master, I have an issue where the dependency graph becomes inconsistent, in the sense that there is a node `u` with a predecessor `v` but `u` is not also stored as the successor of `v` or vice versa. I think this happens when we call `replaceNode` where `replacement` is already a node in the graph. Is this even supposed to happen?

https://github.com/LoAT-developers/LoAT/blob/e4228bbff8b82d4254aa6924927c373996226688/src/its/dependencygraph.hpp#L85-L105

For example, let's say there are two nodes `1` and `2` which form a cycle. Then the successor/predecessor maps look like this:
```
successors = { 1 -> 2, 2 -> 1 }
predecessors = { 1 -> 2, 2 -> 1 }
```
If we now call `replaceNode(1, 2)` then I would expect the updated to maps look like this: 
```
successors = { 2 -> 2 }
predecessors = { 2 -> 2 }
```
However, I think this happens:
```cpp

    void replaceNode(Node to_replace, Node replacement) {
        // to_replace = 1, replacement = 2
        // successors = { 1 -> 2, 2 -> 1 }
        // predecessors = { 1 -> 2, 2 -> 1 }
        nodes.insert(replacement); // no-op since replacement is already included
        nodes.erase(to_replace);
        for (const auto &pred: predecessors.at(to_replace)) {
            successors.at(pred).erase(to_replace);
            successors.at(pred).insert(replacement);
            // successors = { 1 -> 2, 2 -> 2 }
            // predecessors = { 1 -> 2, 2 -> 1 }        
        }
        for (const auto &pred: successors.at(to_replace)) {
            if (pred == replacement) {
                predecessors.at(to_replace).erase(to_replace);
                predecessors.at(to_replace).insert(replacement);
                // successors = { 1 -> 2, 2 -> 2 }
                // predecessors = { 1 -> 2, 2 -> 1 }
            } else {
                predecessors.at(pred).erase(to_replace);
                predecessors.at(pred).insert(replacement);
            }
        }
        predecessors.emplace(replacement, predecessors.at(to_replace)); // no-op because `replacement` is already contained
        successors.emplace(replacement, successors.at(to_replace)); // no-op because `replacement` is already contained
        predecessors.erase(to_replace);
        successors.erase(to_replace);
        // successors = { 2 -> 2 }
        // predecessors = { 2 -> 1 }
    }
```

So what would be the correct behavior?
1. logic error?
2. `replacement` gets all edges from `to_replace` but the previous edges are removed?
3. `replacement` gets all edges from `to_replace` additional to the previous edges?

**Comment by ffrohn** [comment]
I think you're right (actually, I encountered the same issue in my tests, but I haven't had time to look into it). Solution 3. is the right one. I tried to [fix it](https://github.com/LoAT-developers/LoAT/commit/153739742c9a51c4d6428efc202d854910c4d4f9), but I only tested in on a single example so far.
